crates/sc-consensus-subspace-rpc/src/lib.rs

```diff
@@ -828,31 +851,108 @@ where
+                                - MAX_OBJECT_MAPPINGS_IN_CACHE;
+                            if let Some(new_first_item) = cached_archived_object_mappings
+                                .iter()
+                                .nth(new_first_item)
+                                .cloned()
+                            {
+                                *cached_archived_object_mappings =
+                                    cached_archived_object_mappings.split_off(&new_first_item);
+                            }
+                        }
+                    }
+                }
 
-                stream::iter(objects)
+                // Just return the piece indexes in the subscription response
+                stream::iter(objects.into_iter().map(|mapping| mapping.piece_index))
             })
-            .ready_chunks(OBJECT_MAPPING_BATCH_SIZE)
-            .map(|objects| GlobalObjectMapping::V0 { objects });
+            .ready_chunks(MAX_OBJECT_MAPPING_INDEX_BATCH);
 
         self.subscription_executor.spawn(
             "subspace-archived-object-mappings-subscription",
             Some("rpc"),
             pipe_from_stream(pending, mapping_stream).boxed(),
         );
     }
+
+    // Note: this RPC uses the cached object mappings, which are only updated by archived object mappings subscriptions
+    async fn recent_archived_object_mappings(
+        &self,
+        piece_index_start: PieceIndex,
+        limit: PieceIndex,
+    ) -> Result<GlobalObjectMapping, Error> {
+        self.deny_unsafe.check_if_safe()?;
+
+        // There can be zero or many mappings per piece, so this check is only provides quick
+        // rejection of large requests.
+        if u64::from(limit) as usize > MAX_OBJECT_MAPPINGS_PER_REQUEST {
+            error!(
+                "Request limit ({}) exceed the server limit: {} ",
+                limit, MAX_OBJECT_MAPPINGS_PER_REQUEST
+            );
+
+            return Err(Error::StringError(format!(
+                "Request limit ({}) exceed the server limit: {} ",
+                limit, MAX_OBJECT_MAPPINGS_PER_REQUEST
+            )));
+        };
+
+        // Clamp the request if it exceeds the maximum piece index.
+        let piece_index_end = u64::from(piece_index_start)
+            .checked_add(limit.into())
+            .unwrap_or(u64::MAX);
+        let piece_index_end = GlobalObjectArchiveOrder::range_bound(piece_index_end.into());
+        let piece_index_start = GlobalObjectArchiveOrder::range_bound(piece_index_start);
+
+        // We can't tell the difference between requests for mappings that don't exist, and
+        // requests for pruned past cached mappings, or potential future mappings. So we just
+        // return an empty or partly empty response.
+        let objects = self
+            .cached_archived_object_mappings
+            .lock()
+            .range(piece_index_start..piece_index_end)
```

Comment:
What should happen if the number of mappings in this range is larger than `MAX_OBJECT_MAPPINGS_PER_REQUEST`?

If we return an error, then the caller doesn't know how big to make the next request. If we truncate the response, then the caller doesn't know any mappings are missing. (Unless they compare against the subscription piece indexes.)